Feature health armor

yoavk-sudo opened the pull request:
Made class RegeneratingStats : Stat, from which Health and Armor inherit.
Made tests for Health and Armor classes

yoavk-sudo committed:
Finished health tests prior to developing

yoavk-sudo committed:
Finished writing tests for health

yoavk-sudo committed:
Started armor tests

yoavk-sudo committed:
Started armor tests

yoavk-sudo committed:
Added regen class, almost done with armor and health

yoavk-sudo committed:
Ready for review

GDIVX commented on Assets/Scripts/Stats/Health.cs:
```diff
@@ -1,23 +1,31 @@
 {
-    [SerializeField] bool _isInvincible;
+    [SerializeField] float _regenRateInSeconds;
+    [SerializeField] Armor _armor;
```
This creates strong coupling. Armor and health are separate components and thus should remain independent of each other. 
I suggest having an interface IDamageable that would be responsible for dealing with taking damage, and from there have a class DamageHandler that would implement the interface. From there, force me to use health and allow me to choose if the entity has armor or not.

yoavk-sudo replied:
There is a new commit, implementing the interface IDamageable and the DamageHandler class

GDIVX commented on Assets/Scripts/Stats/Health.cs:
```diff
@@ -1,23 +1,31 @@
+    [SerializeField] float _regenRateInSeconds;
+    [SerializeField] Armor _armor;
+    [SerializeField, ReadOnly] bool _isInvincible;
+    [SerializeField, ReadOnly] bool _isDead;
```
Would be more efficient to use a property 
`IsDead => health <= 0`
Otherwise, if your goal was to allow me to kill an entity from the inspector, it shouldn't be read-only.

GDIVX commented on Assets/Scripts/Stats/Health.cs:
```diff
@@ -26,24 +34,64 @@ public override void OnValueChange()
+            if (Value < MaxValue) Value++;
+            yield return new WaitForSeconds(_regenRateInSeconds);
+        }
+        Debug.Log("finished regenerating!");
```
remove debug logs after use

GDIVX commented on Assets/Scripts/Stats/Armor.cs:
```diff
@@ -0,0 +1,45 @@
+            if (Value < MaxValue) Value++;
+            yield return new WaitForSeconds(RegenRateInSeconds);
+        }
+        Debug.Log("finished regenerating!");
```
please remove debug logs after use

GDIVX reviewed (changes requested):
Very well done! Please go over my comments and make a few minor changes.

yoavk-sudo committed:
Implemented IDamageable interface and DamageHanlder class

yoavk-sudo committed:
Changed array of damageables to list and added comment.

Also changed TakeDamage method in DamageHandler to HandleTakingDamage to avoid confusion with the IDamageable method of the same name

yoavk-sudo committed:
monobehaviour on  damageable

yoavk-sudo committed:
Removed List of IDamageables and now using Armor and Health through the inspector
